_rules/text-node-contrast-afw4f7.md

```diff
@@ -0,0 +1,345 @@
+
+## Background
+
+- [Understanding Success Criterion 1.4.3: Contrast (Minimum)](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html)
+- [G18: Ensuring that a contrast ratio of at least 4.5:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G18)
+- [G145: Ensuring that a contrast ratio of at least 3:1 exists between text (and images of text) and background behind the text](https://www.w3.org/WAI/WCAG21/Techniques/general/G145)
+- [F83: Failure of Success Criterion 1.4.3 and 1.4.6 due to using background images that do not provide sufficient contrast with foreground text (or images of text)](https://www.w3.org/WAI/WCAG21/Techniques/failures/F83)
+- [CSS Scoping Module Level 1 (Editor's Draft)](https://drafts.csswg.org/css-scoping/)
+
+## Test Cases
+
+### Passed
+
+#### Passed Example 1
+
+This black [text node](https://dom.spec.whatwg.org/#text) is on a white background.
+
+```html
+<p style="color: #333; background: #FFF;">
+	Some text in a human language
+</p>
+```
+
+#### Passed Example 2
+
+This black [text node](https://dom.spec.whatwg.org/#text) is on a partially white gradient background.
+
+```html
+<p style="color: #333; background: linear-gradient(to right, #FFF, #00F);">
+	Some text in a human language
+</p>
+```
+
+#### Passed Example 3
+
+This light [text node](https://dom.spec.whatwg.org/#text) is on a partially black background image.
+
+```html
+<p
+	style="color: #CCC; height:50px; padding-top:25px; background: #000 no-repeat -20px -20px url('../test-assets/contrast/black-hole.jpeg');"
+>
+	Black hole sun
+</p>
+```
+
+#### Passed Example 4
+
+This [text node](https://dom.spec.whatwg.org/#text) passes because of a text shadow.
+
+```html
+<p style="color: #000; background: #737373; text-shadow: white 0 0 3px">
+	Some text in a human language
+</p>
+```
+
+#### Passed Example 5
+
+This text only has a 3:1 contrast but it is a large (18 point) text so it still passes.
+
+```html
+<p style="color: #000; font-size:18pt; background: #666;">
+	Some text in a human language
+</p>
+```
+
+#### Passed Example 6
+
+This text only has a 3:1 contrast but it is a large (14 points and bold) text so it still passes.
+
+```html
+<p style="color: #000; font-size:14pt; font-weight:700; background: #666;">
+	Some text in English
+</p>
+```
+
+#### Passed Example 7
+
+This text has an insufficient contrast but it is decorative so it still passes.
+
+```html
+<p>Helvetica is a widely used sans-serif typeface developed in 1957 by Max Miedinger and Eduard Hoffmann.</p>
+<p style="font-family: helvetica; background: #EEE; color: #777;" aria-hidden="true">
+	The quick brown fox jumps over the lazy dog.
+</p>
+```
+
+#### Passed Example 8
+
+This text has an insufficient contrast but it does not express anything in human language so it still passes.
+
+```html
+<p style="color: #000; background: #666;">
+	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@
```

Comment:
I'm still challenging a bit the definition of "human language".
I think something like that would fail:
```html
Can you decipher the secret code here?
<p style="color: #000; background: #666;">
	----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@
</p>
```
The point here being that this still doesn't really express anything in a human language, but it is important that it has good contrast. Additionally, the same element could pass or fail depending on context.

As far as I can tell, WCAG says nothing about human language but has "For example, if random words are used to create a background and the words could be rearranged or substituted without changing meaning". Would "words (or charcters?) can be rearranged without altering meaning" be a better formulation of the exception? (that would pass the garbage alone but fail it if part of a cipher (unless this is an elaborate cipher where characters order doesn't matter...))

Reply:
I would state that a secret code is definitly a human language.

In a different context this specific element would indeed fail. But because that specific context is missing in this example we pass it.

Reply:
If context is important and elements cannot be checked individually in isolation, we should probably add a note about it.

Reply:
Strictly speaking, as "----=====++++++++___________***********%%%%%%%%%%%±±±±@@@@@@@@" is not written in a human language, this would not be considered text according to WCAG.

"Text: sequence of characters that can be programmatically determined, where the sequence is expressing something in human language".

I am aware that the definition of text differs from "text node", which is used in this rule. However, I would suggest moving this test case to inapplicable,

Reply:
Agree with Dagfinn. This isn't human language, even if it's a code. It may require a text alternative (unless it's a test, see 1.1.1). There is no scenario in which 1.4.3 applies, but 1.4.11 might in case of a code.

Reply:
I still feel that the definition of Human Language (in WCAG) is open-ended enough to allow cipher ("_language that is spoken, written or signed (through visual or tactile means) to communicate with humans_") and that it is not clear when a secret code becomes a human language or not (for example, pig latin or Templar cipher are widely known and I would say they are human languages, otoh it's not clear that what's written on Phaistos Disc is Human language even though it might have been super clear for people of that time…)

But that's OK, I'm happy with the discussion and the explanation and won't push it further 😄